File: general/depth-first search/0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        boolean[] visited= new boolean[numCourses];
        boolean[] inRecursion= new boolean[numCourses];
        for(int i=0; i<numCourses; i++){
            adj.add(new ArrayList<>());
        }

        for(int[] prereq: prerequisites){
            int u= prereq[1];
            int v= prereq[0];

            adj.get(u).add(v);
        }

        for(int i=0; i<numCourses; i++){
            if(!visited[i] && hasCycle(adj,i,visited,inRecursion)){
                return false;
            }
        }

    return true;
    }

    public boolean hasCycle(ArrayList<ArrayList<Integer>> adj, int u, boolean[] visited,boolean[] inRecursion ){
        visited[u]=true;
        inRecursion[u]=true;

        for(int v:adj.get(u)){
            if(!visited[v]){
                if(hasCycle(adj, v, visited, inRecursion)){
                    return true;
                }
            }else if(inRecursion[v]){
                    return true;
            }
        }

        inRecursion[u]=false;

        return false;
    }
}

File: tesla/depth-first search/0207-course-schedule/0207-course-schedule.ts
function canFinish(numCourses: number, prerequisites: number[][]): boolean {
    const adj = Array.from({length: numCourses}, ()=>[])
    let visited: number[] = Array.from({length:numCourses},()=>0);
// visited = 2, visiting= 1, unviistei =0;

    console.log(visited); 

    for(let [course, prereq] of prerequisites){
        adj[prereq].push(course);
    }   

    for(let i =0; i<numCourses; i++){
        if(dfs(adj,i,visited)){
            return false;
        }
    }

    return true;
};

function dfs(adj: number[][],i: number, visited:number[] ): boolean {
// If node is in current path → cycle
    if (visited[i] === 1) return true;

    // If already processed → no cycle
    if (visited[i] === 2) return false;

    // Mark as visiting
    visited[i] = 1;

    // Explore neighbors
    for (let neighbor of adj[i]) {
        if (dfs(adj, neighbor, visited)) {
            return true;
        }
    }

    // Mark as done
    visited[i] = 2;

    return false;
   
}